dgc/Sign1CoseMessage.cs

```diff
@@ -89,181 +85,95 @@ public byte[] EncodeToBytes()
                 protectedMap[HeaderKey_Alg] = Alg_PS256;
             }
 
-            coseObj.Add(protectedMap.EncodeToBytes());
+            var kidBytes = Convert.FromBase64String(KID);
+            protectedMap[HeaderKey_KID] = CBORObject.FromObject(kidBytes);
 
+            coseObj.Add(protectedMap.EncodeToBytes());
+            
             var unProtectedMap = CBORObject.NewMap();
-            var kidBytes = Convert.FromBase64String(KID);
-            unProtectedMap[HeaderKey_KID] = CBORObject.FromObject(kidBytes);
             coseObj.Add(unProtectedMap);
 
-            coseObj.Add(CBORObject.FromObjectAndTag(Content, CoseHeader_Content));
-            coseObj.Add(CBORObject.FromObjectAndTag(Signature, CoseHeader_Signature));
+            coseObj.Add(Content);
+            coseObj.Add(Signature);
+
             return CBORObject.FromObjectAndTag(coseObj, Sign1Tag).EncodeToBytes();
         }
 
-        public void Sign(AsymmetricCipherKeyPair keypair, DGCertSupportedAlgorithm alg, string keyid)
+        public void Sign(AsymmetricAlgorithm key, string keyid)
         {
-            ISigner signer;
-            var signLength = 0;
             var protectedMap = CBORObject.NewMap();
-            if (alg == DGCertSupportedAlgorithm.ES256)
+
+            if (key is ECDsa)
```

Comment:
switch() :-)

Reply:
`switch` is one option - cleaner would be factory (one switch only) and sub-classing where the key type-specific operations are defined

Reply:
Yup, subclassing is good too, cleaner but a larger change.

Reply:
Personally, I find using if statement more readable and cleaner than switch statements if the cases are few.
I really do not like subclassing.